### ISSUE TITLE
TimeoutError treated as deadConnection

### ISSUE THREAD

**Issue by orinciog** [elastic/elasticsearch-js#1271]
## 🐛 Bug Report
Due to https://github.com/elastic/elasticsearch-js/blob/629894adba28b6aaf388ba2243fe9eef4bb6a5fc/lib/Transport.js#L160, each timeout from elasticsearch is treated as deadConnection and this goes very quickly to have a "No Living Connection", even if there are alive connections.

## To Reproduce
Usecase1: A cluster with three nodes and only primary shards, one node goes down(A), the others two are alive(B and C). One index(I1) is splitted on A, B, C and two indices (I2 and I3) are splitted only to B and C.  

From time to time, there are index requests for I1. These requests are receiving Timeout (one shard down - the one from A), but due to that broken line, the connections to B and C are marked as down (retry is 3) until next resurrection and so, all good requests from this interval(deadConnection - next resurrection) for I2 and I3 are not sent to cluster.

es client is initialized with all three nodes.

Usecase2:
- number of hosts < num_retry
- one request (no matter what type) that receives timeout (no matter what reason).
- all following requests receive NoLivingConnection

## Expected behavior
 Usecase1:
 - Requests with data for I1 fail with TimeoutError
 - Requests with data for I2 and I3 are sent to cluster.
 
Usecase2:
 - First request should fail with Timeout (now fails with NoLivingConnection)
 - Following requests should succeed (now they fail with NoLivingConnection)

IMHO, that line should be changed to 
`if (err.name!=="TimeoutError"){
          this.connectionPool.markDead(meta.connection)
        }`
(Basically, a timeout connection is not a dead connection)

## Your Environment

- *node version*: 12
- `@elastic/elasticsearch` *version*: This affects 7.x and 6.x branches.

**Comment by delvedor** [comment]
Hello! If you are getting the `NoLivingConnectionError` it means that you have probably set a too strict filter or selector, as from https://github.com/elastic/elasticsearch-js/pull/1127 the connection pool always return a connection even if there are no alive connections.

The `TimeoutError` should not be confused with the HTTP Timeout (which is a `ResponseError` with `statusCode` equal to `408`). The `TimeoutError` is triggered when the underlying socket goes into timeout (by default `30000ms`).

Which version of the client are you using?

**Comment by orinciog** [comment]
> Hello! If you are getting the `NoLivingConnectionError` it means that you have probably set a too strict filter or selector, as from #1127 the connection pool always return a connection even if there are no alive connections.

No, I don't have any filter or selector, just the defaults. I can provide a valid script with few lines. 

```
const { Client } = require("@elastic/elasticsearch");

const client = new Client({
    node:'http://localhost:9200'
});


async function main(){
    try{
        const r=await client.cat.nodes({},{requestTimeout:1});
        console.log(r);
    }catch(err){
        //NoLivingConnection instead of TimeoutError
        console.error(err);
    }
    try{
        const r=await client.cat.nodes({},{requestTimeout:30000}); 
       //this should work, but it throws NoLivingConnection 
        console.log(r);
    }catch(err){
        //NoLivingConnection 
        console.error(err);
    }
}

main();
```

> Which version of the client are you using?

Latest of 6.x (6.8.7)

**Comment by delvedor** [comment]
#1127 has not been backported to `6.x`, so you are still getting that error.
Have you tried to curl Elasticsearch? If you are getting the `TimeoutError` even with 30s there might be something wrong with the network.

**Comment by orinciog** [comment]
> #1127 has not been backported to `6.x`, so you are still getting that error.
> Have you tried to curl Elasticsearch? If you are getting the `TimeoutError` even with 30s there might be something wrong with the network.

Elasticsearch is on localhost. (I've updated the example, first request is fine, second throws NoLivingConnection, third request (is the same with the first) throws  NoLivingConnection) .

**Comment by delvedor** [comment]
The behavior is expected, in the second request you are setting a timeout of 1 millisecond, which will always timeout, and since the max retries is 3, then all of your nodes will be retried and marked as dead.

**Comment by orinciog** [comment]
Thank you for your reply. 
> The behavior is expected, in the second request you are setting a timeout of 1 millisecond, which will always timeout, and since the max retries is 3, then all of your nodes will be retried and marked as dead, causing the `NoLivingConnectionError`.

Ok, but I think that the third request is not expected behaviour. I think it should response just as the first one, no?

**Comment by delvedor** [comment]
> I think it should response just as the first one, no?

Nope, because in the second request all the nodes have been marked as dead. (it's not *per request*).

> even if all connections are fine.

If in the second request you got a `TimeoutError` for each retry, it means that none of the connections are fine :)

Which version of Elasticsearch are you using?

**Comment by orinciog** [comment]
> Nope, because in the second request all the nodes have been marked as dead. (it's not per request).

You should write this in documentation. 

> If in the second request you got a `TimeoutError` for each retry, it means that none of the connections are fine :)

Sorry but this is wrong. In real world, some requests can always fail, no matter what is the state of network. (See usecase1). Other example: one heavy aggregation. So what you are telling here is that if a TimeoutError occurs, we should deny all requests for one minute (resurrectionTimeout), no?

> Which version of Elasticsearch are you using?
6.2.3

**Comment by delvedor** [comment]
> Sorry but this is wrong. In real world, some requests can always fail, no matter what is the state of network. 

Sure, but the handling depends on the failure. If a request fails with a response error (4xx) is different from a connection error.

> Other example: one heavy aggregation.

If you know it will be a heavy aggregation, you should increase the default timeout, it can be done globally or per request basis.

> So what you are telling here is that if a TimeoutError occurs, we should deny all requests for one minute (resurrectionTimeout), no?

What I'm saying is that if a request has consumed all the available connections, and all of them failed with a timeout error, it is safe to assume that also the subsequent request will fail as well.


I'll release #1127 in 6.x as well, so you should no longer see `NoLivingConnectionError`, but be aware that the root issue is in another place, this fix will only hide it from you.

**Comment by delvedor** [comment]
Landed in `v6.8.8`. Closing the issue, feel free to reopen if you are still encountering the problem.
